### ISSUE TITLE
export types in package.json // Seperate project for bootstrap-icons

### ISSUE THREAD

**Issue by boindil** [cdmoro/bootstrap-vue-3#750]
**Is your feature request related to a problem? Please describe.**
I would like to create a new project bootstrap-vue-3-icons as suggested in #255, however the icons use types from bootstrap-vue-3.

**Describe the solution you would like**
I would prefer not to duplicate these types and instead import the original ones from the bootstrap-vue-3 library.

**Describe alternatives you have considered**
Duplicating the types.

**Additional context**
The affected types are: `Animation`, `ColorVariant`, `TextColorVariant` - `Animation` however seems to be only used by the icons.
Are there any other recommendations you can think of / a better way to do this?

**Comment by VividLemon** [comment]
Importing types has no effect on package size. You could safely import the type from the root. But I may do your suggestion as well

**Comment by boindil** [comment]
Hi @VividLemon,

I provided the seperated everything into https://github.com/boindil/bootstrap-vue-3-icons which is essentially based on the state that the project was in when the icons were seperated. There actually isn't done much - simply grabbing the project, the deleted files and bringing everything together.

I did not publish this to npmjs and would rather like this to get put into one of your workspaces to be officially supported. If there are suggestions on what to change to make this happen more quickly, please tell.

**Comment by seaChris** [comment]
> Importing types has no effect on package size. You could safely import the type from the root. But I may do your suggestion as well. If your system becomes well enough, it could also be put in one of our workspaces and officially supported. I was always planning on making a package for it, directly after a v1.0 release.

Iam very intrested in this.

**Comment by aceofwings** [comment]
@boindil I think it would be great to have this functionality back. @jclijmans contributions were removed some time ago however I think it is important to have. Does tree shake work effectively in your project currently? The original bootstrap vue compiles SVGs and Icon definitions into their own files thus minimizing size based off of usage.

**Comment by boindil** [comment]
@aceofwings as of now I dont think it does, but I can have a look into it.

My next step however will be to convert the project structure to the new one of this project.

**Comment by aceofwings** [comment]
Yeah it will pull the whole svg package. If you need any assistance on structure let me know. I can also work on it as I build the spin button.

**Comment by boindil** [comment]
I actually should be done converting the repository structure. There might be improvements to be done on the actual components as they look quite different in comparison (interfaces, props, ...) but for now it should be fine.

I've already looked into what bootstrap-vue originally did and will try to recreate that mechanic to allow tree-shaking.

If I need any assistance, I will gladly accept your help, but for now I'd like to try and figure it out myself if I can :)

**Comment by VividLemon** [comment]
> If I need any assistance, I will gladly accept your help, but for now I'd like to try and figure it out myself if I can :)

Instead of focusing on the setup of the library, you can create the package in bv3, and I can get everything configured. If you would like, I can take the relevant changes from your library, then you can work on building it in this library.

**Comment by boindil** [comment]
> > If I need any assistance, I will gladly accept your help, but for now I'd like to try and figure it out myself if I can :)
> 
> Instead of focusing on the setup of the library, you can create the package in bv3, and I can get everything configured. If you would like, I can take the relevant changes from your library, then you can work on building it in this library. I will probably work on converting the render functions from BV as well since they're fairly small.

Yes, that sounds great to me. Since the project structure is now matching the one of bv3 this should be an easy one. For now I have tested publishing to npm (@boindil/bootstrap-vue-3-icons) and using it in another project. Works fine so far.

I'll probably start poking around the separate icon-components to allow tree-shaking on monday.

**Comment by VividLemon** [comment]
I'll work on bringing the library into ours soon. Since I don't have access to the repository secrets, I can't set up npm publishing at the moment.

**Comment by VividLemon** [comment]
> Yes, that sounds great to me. Since the project structure is now matching the one of bv3 this should be an easy one. For now I have tested publishing to npm (@boindil/bootstrap-vue-3-icons) and using it in another project. Works fine so far.

I have pulled your library into ours [here](https://github.com/cdmoro/bootstrap-vue-3/tree/main/packages/bootstrap-vue-3-icons). I think the best option for development on your end is probably to create a fork of this library, and become a contributor upstream here. You can read the contributing.md at the root for information if you need it, and are free to ask any questions here on how to help.

I made some changes to what you had, but nothing major. Mostly some package.json changes, added/removed some dependencies. 

I will likely make a few notable changes soon.

1. Rebuild the components to use template and script setup syntax
2. Your original idea was to export types from bootstrap-vue-3... This can still be done, but for the time being I was planning on creating a separate "shared" workspace, which will export shared types and util methods between the two packages internally and not duplicate code, one of the perks of having the workspace.
3. Write tests.

This will also prove as a test project for maintaining multiple workspace packages versions/npm publishes/releases using release please. The release-please github action and bot will need to be modified to allow for multiple releases in a monorepo. This will come in handy to have experience and be set up for in the future when bootstrap-vue-3 and bootstrap-vue are merged. 

Eventually, it would be nice to have control over the npm package, to not have duplicate packages. However, since bootstrap-vue will soon to have Vue 3 compatibility, "bootstrap-vue-3" is a bit odd. I still need to talk with Xanf and Pi0, but I think it will likely be renamed to bootstrap-vue-next, since icons will soon be fully supported once all the issues are worked out, it will likely also be renamed to bootstrap-vue-icons-next

Thank you for getting the project started.

**Comment by boindil** [comment]
Sounds great to me. As for the non package: I intentionally used `@boindil` as prefix. Once everything is separately available from this project, I will deprecate that and hint to the official one.

**Comment by boindil** [comment]
@VividLemon I added a new commit here that hopefully should be close to what we are trying to achieve: https://github.com/boindil/bootstrap-vue-3/commit/6c6ce694ff4ce0313d7f25e79221940bce95f862

I tried to make it close to the original attempt of BV but use the newer structure of this project.

However there are still some things that should or have to be addressed:
- Generation of *.d.ts files fail because there are too many icons in the newest bootstrap-icons version. I used export default just as its done in the BV3 package and get 
```
bootstrap-vue-3-icons:build: [vite:dts] Start generate declaration files...
bootstrap-vue-3-icons:build: src/components/index.ts:15:1 - error TS7056: The inferred type of this node exceeds the maximum length the compiler will serialize. An explicit type annotation is needed.
```
- In the original implementation the specific icon-component might be already loaded by a parent and if so is taken from there. I did not find a way to achieve this using the Vue3 composition API.
- The components `BIcon` and `BIconstack` use the type `ColorVariant` to define the allowed values of the `variant` prop. The component `BIconBase` however uses `TextColorVariant` for some reason. When trying these additional values, I did not see any change. It should be probably shifted towards one or the other and if `TextColorVariant` is correct, it should at least work.
- I did intentionally not include the generation script to be run on every build, because I don't want to interfere there.

Maybe you can have a look and help me out at this point?

**Comment by boindil** [comment]
@VividLemon I guess this has probably slipped through, but would you mind having a look at this and provide me with some feedback?

**Comment by VividLemon** [comment]
> @VividLemon I guess this has probably slipped through, but would you mind having a look at this and provide me with some feedback?

What issue are you referring to?

**Comment by boindil** [comment]
@VividLemon in my last post here, right before this, I posted a link to the commit in my fork where I made the changes.

Since there are still some issues where I am unsure of how to solve them, I asked for your help to get this finished hopefully soon.
- `*.d.ts` generation fails -> named exports instead?
- getting the components off the parent like in Vue2 -> no clue, didnt find anything alike in Vue3 to use
- `ColorVariant` vs `TextColorVariant` on the components

I create a PR as well and mark it as WIP and let you change stuff right there if you'd like that better?
